packages/dev/core/src/Meshes/greasedLineMesh.ts

```diff
@@ -0,0 +1,535 @@
+    }
+
+    private _preprocess(positions: number[]) {
+        const l = positions.length / 6;
+
+        let v: number[] = [];
+
+        const previous = [];
+        const next = [];
+        const side = [];
+        const uvs = [];
+
+        if (GreasedLineMesh._CompareV3(0, l - 1, positions)) {
+            v = GreasedLineMesh._CopyV3(l - 2, positions);
+        } else {
+            v = GreasedLineMesh._CopyV3(0, positions);
+        }
+        previous.push(v[0], v[1], v[2]);
+        previous.push(v[0], v[1], v[2]);
+
+        for (let j = 0; j < l; j++) {
+            side.push(1);
+            side.push(-1);
+
+            // uvs
+            uvs.push(j / (l - 1), 0);
+            uvs.push(j / (l - 1), 1);
+
+            if (j < l - 1) {
+                v = GreasedLineMesh._CopyV3(j, positions);
+                previous.push(v[0], v[1], v[2]);
+                previous.push(v[0], v[1], v[2]);
+            }
+            if (j > 0) {
+                v = GreasedLineMesh._CopyV3(j, positions);
+                next.push(v[0], v[1], v[2]);
+                next.push(v[0], v[1], v[2]);
+            }
+        }
+
+        if (GreasedLineMesh._CompareV3(l - 1, 0, positions)) {
+            v = GreasedLineMesh._CopyV3(1, positions);
+        } else {
+            v = GreasedLineMesh._CopyV3(l - 1, positions);
+        }
+        next.push(v[0], v[1], v[2]);
+        next.push(v[0], v[1], v[2]);
+
+        return {
+            previous,
+            next,
+            uvs,
+            side,
+        };
+    }
+
+    private _drawLine() {
+        const vertexData = new VertexData();
+        vertexData.positions = this._vertexPositions;
+        vertexData.indices = this._indices;
+        vertexData.uvs = this._uvs;
+        vertexData.applyToMesh(this, false);
+
+        const engine = this._scene.getEngine();
+
+        const previousBuffer = new Buffer(engine, this._previous, false, 3);
+        this.setVerticesBuffer(previousBuffer.createVertexBuffer("previous", 0, 3));
+
+        const nextBuffer = new Buffer(engine, this._next, false, 3);
+        this.setVerticesBuffer(nextBuffer.createVertexBuffer("next", 0, 3));
+
+        const sideBuffer = new Buffer(engine, this._side, false, 1);
+        this.setVerticesBuffer(sideBuffer.createVertexBuffer("side", 0, 1));
+
+        const countersBuffer = new Buffer(engine, this._counters, false, 1);
+        this.setVerticesBuffer(countersBuffer.createVertexBuffer("counters", 0, 1));
+
+        const widthBuffer = new Buffer(engine, this._widths, this._updatable, 1);
+        this.setVerticesBuffer(widthBuffer.createVertexBuffer("widths", 0, 1));
+        this._widthsBuffer = widthBuffer;
+
+        if (this._offsets) {
+            const offsetBuffer = new Buffer(engine, this._offsets, this._updatable, 3);
+            this.setVerticesBuffer(offsetBuffer.createVertexBuffer("offsets", 0, 3));
+            this._offsetsBuffer = offsetBuffer;
+        }
+    }

```

Comment:
See my comment in the plugin material, we should try to pack some of the data to reduce the number of attributes (so, pack side + counters + widths if possible).

Reply:
Didn't read/hear anything about packing vertex buffers until now. I did some googling and if I understand the articles I've read correctly I can use for example a vec3 for packing three ints and access them in the shader as data.x, data.y, data.z. I've read a bit about vertex data layouts. Which technique is preffered?

Reply:
So finally I didn't pack side + counters + widths because widths can be updatable. I could pack only side + counters and end up wioth 4 buffers but better solution was to pack previousAndSide and nextAndCounters to vec4 so I saved one more buffer resulting to final count of 3 from 5 before (+ offsets buffer if used).